\subsection*{Traditional Genetic Operators}
Genetic Algorithms were designed to emulate evolution by natural selection, and traditional genetic operators closely resemble their biological counterparts. Crossover, mutation, and selection build the core of many Genetic Algorithms, and are direct translations of biological processes into the computational world. \cite{Russell10}. For the rest of this paper, the term ``traditional Genetic Algorithm'' will be used to refer to 1-point crossover and mutation together.

\subsubsection*{1-Point Crossover}
1-point crossover is the computational analog to sexual reproduction. This binary genetic operator takes two genetic sequences, and then chooses a splitting point in their genetic sequences. In the example below, the splitting point is between the fourth and fifth alleles:

\begin{align*}
\text{Parent 1} &: {\color{myred}0~1~1~0~}|{\color{myred}~1~0~1~1~0} 			\\
\text{Parent 2} &: {\color{myblue}0~1~0~0~}|{\color{myblue}~0~0~1~0~1} 	\\		
\end{align*}

\noindent The alleles that occur after the splitting point are then swapped between the two parents to create two new child genetic sequences. The parents from the example above would produce the children below:

\begin{align*}
\text{Child 1} &: {\color{myred}0~1~1~0}{\color{myblue}~0~0~1~0~1} 			\\  
\text{Child 2} &: {\color{myblue}0~1~0~0}{\color{myred}~1~0~1~1~0} 		
\end{align*}

\subsubsection*{Uniform Mutation}
Mutation is a unary genetic operator designed to help Genetic Algorithms break out of local optima \cite{Russell10}. This operator scans through a genetic sequence of length $n$, and at each point will change that allele to another value in that position's domain with a given probability $p$. This probability is typically set to $\frac{1}{n}$, but further research has demonstrated that a dynamically set $p$ produces better behavior \cite{Back93}. An example has been provided below:

\begin{align*}
\text{Before Mutation} &: 0~1~1~0~0~0~1~0~1 			\\  
\text{After Mutation} &: 0~1~1~{\color{myred}1}~0~0~1~0~1  		
\end{align*}

\subsubsection*{Swapping Mutation}
Problems that require their genetic sequences to be permutations, like the Traveling Salesman Problem, utilize swapping, instead of uniform, mutation. Both operators scan through their respective genetic sequences, but the difference occurs when a change is made in the case of permutations. These mutation operators choose a random point in the rest of the genetic sequence and swap the values at these two locations. An example can be seen below:

\begin{align*}
\text{Before Mutation} &: 8~4~2~1~9~7~3~5~6 			\\
\text{After Mutation} &:  8~{\color{myred}7}~2~1~9~{\color{myred}4}~3~5~6		
\end{align*}

\subsubsection*{Steady-State Selection}
Steady-state selection is used to ensure that underperforming genetic sequences are eliminated from the population, thus preventing their genes from propagating further. Once recombination and mutation have finished adding individuals to the population for a particular generation, this operator assigns each individual a fitness value with the fitness function for the current problem. The individuals are then sorted in descending order with respect to their fitness values. The first $k$ individuals in this list become the population operated upon in the next generation.

\subsection*{Multi-Parent Genetic Operators}
The literature concerning multi-parent genetic operators highlighted two central operators: fitness-based scanning and diagonal crossover. These two strategies have shown success as multi-parent genetic operators, and are the direct conceptual descendants of traditional genetic operators. Both of these operators will be tested independently, as well as in conjunction with each other. This \emph{mixed-operator} methodology has shown performance gains in previous research \cite{Smith04}.

\subsubsection*{Fitness-Based Scanning}
The first operator from the literature that has been implemented and tested is fitness-based scanning as proposed by Eiben, et al \cite{Eiben94}. The generalized form of scanning iterates through the empty genetic sequence of a child and determines its value based upon the values present in $n$ selected parents \cite{Eiben91}. Fitness-based scanning makes a roulette wheel selection to choose each allele. At each allele, the probability that a parent, from the sub-population, consisting of the parent genetic sequences selected for this recombination, $S$, named $i$ with a fitness value of $f(i)$, will donate its allele to the child's genetic sequence has probability $P(i)$ as described below \cite{Eiben94}: 

\[ P(i) = \frac{f(i)}{\sum\limits_{i \in S} f(i)} \]

\noindent Thus, the expected number of alleles inherited from a parent $i$ is $E(i)$ \cite{Eiben94}:

\[ E(i) = P(i) *(\text{ Chromosome length }) \]

In previous studies, fitness-based scanning had mixed results across several numeric optimization functions, and performed well on the Traveling Salesman Problem \cite{Eiben94, Eiben95}. Overall, fitness-based scanning managed to surpass the results of traditional crossover and other multi-parent genetic operators \cite{Eiben94}. 

\subsubsection*{Diagonal Crossover}
\emph{Diagonal crossover} was introduced by Eiben, et al. to extend the concept of crossover into the realm of multi-parent genetic operators \cite{Eiben03}. A diagonal crossover of arity $n$ can be described easily. Take $n$ individuals from a population, select $n-1$ crossover points, and create $n$ children by selecting one subsequence from each piece of the genetic sequences given \cite{Eiben95}. Figure \ref{DC-Fig} depicts how this would work \cite{Eiben95}. Note that $a_i, b_i, c_i,$ and $d_i$ are genetic subsequences of arbitrary length.
\begin{figure}[h!]
\centering
\begin{tabular}{ | c | c | c | c | c | }
\hline
{\color{myred}$a_1$} & {\color{myred}$a_2$} & {\color{myred}$a_3$} & {\color{myred}$a_4$} & $\text{ parent a }$ 	\\ \hline
{\color{myblue}$b_1$} & {\color{myblue}$b_2$} & {\color{myblue}$b_3$} & {\color{myblue}$b_4$} & $\text{ parent b }$ 	\\ \hline
{\color{mygreen}$c_1$} & {\color{mygreen}$c_2$} & {\color{mygreen}$c_3$} & {\color{mygreen}$c_4$} & $\text{ parent c }$ 	\\ \hline
$d_1$ & $d_2$ & $d_3$ & $d_4$ & $\text{ parent d }$ 	\\ \hline
\end{tabular}
$\rightarrow$
\begin{tabular}{ | c | c | c | c | c | }
\hline
{\color{myred}$a_1$} & $d_2$ & {\color{mygreen}$c_3$} & {\color{myblue}$b_4$} & $\text{ child a }$ 	\\ \hline
{\color{myblue}$b_1$} & {\color{myred}$a_2$} & $d_3$ & {\color{mygreen}$c_4$} & $\text{ child b }$ 	\\ \hline
{\color{mygreen}$c_1$} & {\color{myblue}$b_2$} & {\color{myred}$a_3$} & $d_4$ & $\text{ child c }$ 	\\ \hline
$d_1$ & {\color{mygreen}$c_2$} & {\color{myblue}$b_3$} & {\color{myred}$a_4$} & $\text{ child d }$ 	\\ \hline
\end{tabular}
\caption{Diagonal crossover applied to four parents}
\label{DC-Fig}
\end{figure} 

The rationale behind expanding this operator into the realm of multi-parenthood was to increase the disruptiveness, and by extension the explorativity, of traditional crossover \cite{Eiben97}. This meant that the population would need a large degree of similar genetic sequences before the search would narrow and converge \cite{Eiben95}. It should also be noted that in the special case of $n = 2$ is identical to traditional 1-point crossover \cite{Eiben95}. High arity versions of this operator increased the performance of Genetic Algorithms in research, though they were ultimately less successful than scanning operators; however, it was also noted that it is much less expensive to compute, meaning that larger populations could be processed in the same time \cite{Eiben95}. In our research, the population sizes are set to be equal across all operators.

Now that we have defined each of the genetic operators from the literature, we will define our contribution: the genetic overlay and its concrete implementation, the elitist schema overlay.